### ISSUE TITLE
Enabling `CHECK_PTR_SIZES` in latest build causes fatal asserts

### ISSUE THREAD

**Issue by Tisten** [OurMachinery/themachinery-public#1307]
- **Version:** Git Master Branch
- **OS:** Windows

Please run the-machinery.exe with `CHECK_PTR_SIZES` enabled in `foundation/memory_tracker.c`

The issues I've run into has been in the ui plugin, and have triggered with only the default project opened.

**Comment by Tisten** [comment]
I think moving this scope:
https://github.com/OurMachinery/themachinery/blob/ea20c67a22819be4ec6292e090fdaac623c9a81e/foundation/memory_tracker.c#L317-L321
one line up into the `if (ti)` scope would solve them.

**Comment by Tisten** [comment]
CHECK_PTR_SIZES was added by @niklas-ourmachinery so I guess this bug and https://github.com/OurMachinery/themachinery-public/issues/1306 are for him.

**Comment by simon-ourmachinery** [comment]
Note: CI make sure we run with this setting.

**Comment by Tisten** [comment]
> Note: CI make sure we run with this setting.

I tried running the tests with this, and they were fine. I guess the tests of the ui plugin are a bit limited. The first issue I got was allocation of index buffers.

**Comment by niklas-ourmachinery** [comment]
I spent a fair amount of time digging into this... and I think I've found the issue.

Suppose we have two threads where one thread does:

~~~c
free(p1);
untrace(p1);
~~~

And the other one does:

~~~c
p2 = alloc();
trace(p2);
~~~

And suppose the operations get executed in this order:

~~~c
free(p1);
p2 = alloc();
trace(p2);
untrace(p1);
~~~

The allocation of `p2` might return the same pointer `p1` that was freed by the other thread (since that memory is now freed). In this case, `untrace()` will remove that pointer from the trace and `p1` will be untraced after this, causing the error.

This triggers in the UI, because the UI is multithreaded and allocates a number of pretty large fixed size buffers, so there is a decent chance that the allocator will reuse one that was just freed.

To fix this, `untrace()` must happen before `free()` while the pointer is still "owned" by the thread. Unfortunately, in this code:

~~~
 new_ptr = realloc(old_ptr, new_size);
 tm_memory_tracker_api->record_realloc(old_ptr, old_size, new_ptr, new_size, file, line, a->mem_scope);
 ~~~
 
We can't call `record_realloc()` before we have `new_ptr`... so either we have to change the `tm_memory_tracker_api`, or we can't use system `realloc()`, but must instead use separate `malloc()` and `free()` calls with free being called *after* the call to `record_realloc()`.

I will think this over a bit to see what is the best solution.
